### ISSUE TITLE
argocd app diff --local does not pull in the app's kustomize options

### ISSUE THREAD

**Issue by snuggie12** [argoproj/argo-cd#5532]
If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a question in argocd slack [channel](https://argoproj.github.io/community/join-slack).

Checklist:

* [X] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [ ] I've included steps to reproduce the bug.
* [ ] I've pasted the output of `argocd version`.

**Describe the bug**

When running an `argocd app diff $app_name --local $path` the live version will use the custom version of kustomize specified in the app, but the local one only includes kustomize options from the `ConfigMap` (e.g. `argocd-cm`).

**To Reproduce**

1. Build a custom repo server image with multiple versions of kustomize. If you want a specific example I'd say put 3.9.3 on there.
2. Deploy an app and specify to use the custom version.
3. For my specific example of 3.9.3, make a `ConfigMap` and make one of the keys under `.data` be an integer (e.g. `.data.1234: foo`)
4. On your computer or anywhere with the argocd CLI available make sure `$(which kustomize) version` reports the version of kustomize that the repo server has by default (I think it's 3.8.1 currently in master?)
5. Now a regular sync will break since kyaml is turned on in 3.9.3 but a diff will be fine since 3.8.1 does not have kyaml turned on by default.
6. You can reverse this by taking some other random version (e.g. 3.6.1) and assigning it to your app, but then symlink or cp 3.9.3 where `which kustomize` reports. Now your app will sync fine but your diff will throw json errors.

**Expected behavior**

Kustomize options from the app source and from argo settings should be merged before manifest generation. Since `argocd-cm` kustomize options are loaded during a diff --local that should mean that the binary path option is present. I'd suggest in `getBinaryPath` that a check is done that the file exists. If the file does not exist a warning is printed and a fall back to `"kustomize"` happens (as it is now if no custom binary option is present.)

**Version**

1.7.8 but I believe it's like this in master.

**Comment by alexmt** [comment]
Hello @snuggie12 ,

You are right, the kustomize binary path is ignored when built locally. Agree with your suggestion that we should gracefully handle missing binary locally.

**Comment by chetan-rns** [comment]
Hi @alexmt @snuggie12, I noticed that the binary path is always " " inside `getBinaryPath` when using diff --local. I suspect it's being [ignored here](https://github.com/argoproj/argo-cd/blob/8595c2f0a927e7631dd6fed551f35019112d291a/server/settings/settings.go#L88). Are you suggesting that instead of ignoring, we use the path specified in argocd-cm and check if the binary path actually exists locally?

**Comment by snuggie12** [comment]
I'm not a programmer so take this with a grain of salt, but I thought it was this:
- https://github.com/argoproj/argo-cd/blob/2a1a03a91996c176dabe7b2ee7fe49cd6ac58b62/cmd/argocd/commands/app.go#L823 (Settings are only settings in `argocd-cm`)
- https://github.com/argoproj/argo-cd/blob/2a1a03a91996c176dabe7b2ee7fe49cd6ac58b62/reposerver/repository/repository.go#L651 (We never pulled the app's settings so `kustomize` is what is returned.)

Comparing that to this call done by the controller where the application itself is looked at I think is the problem: https://github.com/argoproj/argo-cd/blob/baa0f2e39c45b86bedc3b3cca0878c0f77253529/controller/state.go#L154

Not sure if possible, but since the live resources are fetched to compare against I would think that's when you could grab the app-specific kustomize options and then pass that along to the local version.

That's the first part of the issue, the other is that there is no check for the binary. This works for the repo server since the default build is guaranteed to have `kustomize` in the `PATH`, but if using `--local` a warning should at least be thrown: https://github.com/argoproj/argo-cd/blob/7f0ffb4cd2184a5ff506686c270eb44a3e0e2b50/util/kustomize/kustomize.go#L53-L58

This could be implemented in a variety of ways and not sure I know what is best. You could have a parameter that is treating `--local` special. You could have a parameter called `actionOnMissing` that warns or errors.

**Comment by chetan-rns** [comment]
@snuggie12  Thanks for the detailed explanation. I may be wrong, but AFAIU, there are no application-specific options

> kustomizeOptions, err := kustomizeSettings.GetOptions(app.Spec.Source) 

The options that we get above in the controller are still coming from the settings in `argocd-cm`. Maybe implementing #5533 could give options that are specific to the kustomize version used by the application.

**Comment by snuggie12** [comment]
We might be talking about different things, but I'm referring to this (from [here](https://argoproj.github.io/argo-cd/user-guide/kustomize/#custom-kustomize-versions) ):

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
spec:
  source:
    repoURL: https://github.com/argoproj/argocd-example-apps.git
    targetRevision: HEAD
    path: guestbook-kustomize

    kustomize:
      version: v3.5.4
```

And `v3.5.4` links back to the `argocd-cm` config map where you specify the path.

**Comment by mikedougherty** [comment]
In addition to the binary location, `kustomize.buildOptions` for the application's specified kustomize version are not fetched by the client.

When the argocd repo-server renders my application manifest (using kustomize v4.1.2, with `kustomize.buildOptions.v4.1.2: --enable-helm` in the `argocd-cm`, it is able to render the manifests correctly.

When I run `argocd diff my-app --local .` from my workstation, it fails with an error stating `must specify --enable-helm`. This seems to indicate that the `buildOptions` are not available to the CLI client.

**Comment by snuggie12** [comment]
Does any of this get fixed with the new focus on config management plugins?

Should I create a new issue for the global versioned options also not running?
